oauth/services/openshift.py

```diff
@@ -0,0 +1,187 @@
+        return OAuthEndpoint(
+            urllib.parse.urlunsplit((scheme, netloc, path, {}, fragment)), query_params
+        )
+
+    @lru_cache(maxsize=1)
+    def _oauth_config(self):
+        if self.config.get("OPENSHIFT_SERVER"):
+            return self._load_oauth_config_via_discovery(self.config.get("DEBUGGING", False))
+        else:
+            return {}
+
+    def _load_oauth_config_via_discovery(self, is_debugging):
+        """
+        Attempts to load the OAuth config via the OAuth discovery mechanism using OpenShift Default Service DNS.
+
+        If is_debugging is True, non-secure connections are alllowed. Raises an
+        DiscoveryFailureException on failure.
+        """
+        oauth_server = self.config.get("OPENSHIFT_SERVER", DEFAULT_OAUTH_HOST)
+        # if not oidc_server.startswith("https://") and not is_debugging:
+        #     raise DiscoveryFailureException("OIDC server must be accessed over SSL")
+
+        discovery_url = urllib.parse.urljoin(oauth_server, OAUTH_WELLKNOWN)
+        discovery = self._http_client.get(discovery_url, timeout=5, verify=is_debugging is False)
+        if discovery.status_code // 100 != 2:
+            logger.debug(
+                "Got %s response for OpenShift OAuth discovery: %s", discovery.status_code, discovery.text
+            )
+            raise DiscoveryFailureException("Could not load OpenShift OAuth discovery information")
+
+        try:
+            return json.loads(discovery.text)
+        except ValueError:
+            logger.exception("Could not parse OpenShift OAuth discovery for url: %s", discovery_url)
+            raise DiscoveryFailureException("Could not parse OpenShift OAuth discovery information")
+
+    def get_public_config(self):
+        return {
+            "CLIENT_ID": self.client_id(),
+            "OAUTH": True,
+        }
+
+    def exchange_code_for_login(self, app_config, http_client, code, redirect_suffix):
+        """
+        Exchanges the given OAuth access code for user information on behalf of a user trying to
+        login or attach their account.
+
+        Raises a OAuthLoginService exception on failure. Returns a tuple consisting of (service_id,
+        service_username, email)
+        """
+
+        # Retrieve the token for the OAuth code.
+        try:
+            token = self.exchange_code_for_token(
+                app_config,
+                http_client,
+                code,
+                redirect_suffix=redirect_suffix,
+                form_encode=self.requires_form_encoding(),
+            )
+        except OAuthExchangeCodeException as oce:
+            raise OAuthLoginException(str(oce))
+
+        # Retrieve the user's information with the token.
+        try:
+            user_info = self.get_user_info(http_client, token)
+        except OAuthGetUserInfoException as oge:
+            raise OAuthLoginException(str(oge))
+
+        # if user_info.get("id", None) is None:
```

Comment:
Is this intended to be commented out?

Reply:
Let me see whether I can actually fetch profile information from the user-info endpoint. Otherwise automatically created Quay users just get the login carried over and nothing else.

Reply:
This one still pending user-info endpoint tests